### ISSUE TITLE
Use CoveringQuery to select percolator candidates?

### ISSUE THREAD

**Issue by jpountz** [elastic/elasticsearch#26307]
Currently when extracting conjunctions such as `a AND b AND c`, we use heuristics in order to figure out which term is the rarest and we only index this term.

Lucene just got a new [`CoveringQuery`](https://issues.apache.org/jira/browse/LUCENE-7621) which allows to configure the number of required clauses on a per-document basis. This means that we could index all terms of `a AND b AND c` alongside a numeric field that would store `3` and then at query time the created `CoveringQuery` wouldn't have false positives: we would not even need to run the query through the memory index to check whether it matches.

**Comment by martijnvg** [comment]
I worked on integrating the [CoveringQuery in the percolator](https://github.com/martijnvg/elasticsearch/commit/c714748e856ba4190e5f7ab5c1f25d73e3c41c0a) and it is a good fit! I tested this change out on a percolator query set and the query time was reduced by 3 times.

I did run into two challenges:
* Currently all the extracted terms from the document to be percolated are being put in a `TermInSetQuery` instance. This is no longer possible with the `CoveringQuery`, because then all terms inside `TermsInSetQuery` count as single match. So currently in the commit mentioned above [each query term is added as a separate`TermQuery` instance](https://github.com/martijnvg/elasticsearch/commit/c714748e856ba4190e5f7ab5c1f25d73e3c41c0a#diff-7a987e4fb3c8a1820bfca74aa3e8b5c0R316) to the `CoveringQuery`. This is not ideal because if a large document is to be percolated then could cause a `TooManyClauses` exception. I need to think about this more on how to solve this.
* In case of range queries each extraction does not simply increment the minimum_should_match for that percolator query like for a term based extraction, so that can lead to more false positives for percolator queries with range queries than term based queries. The is because the way number fields are extracted from the document to be percolated. Per field a single range is extracted and if a percolator query has two or more range queries on the same field than the the minimum should match can be higher than clauses in the CoveringQuery. Therefore right now the minimum should match [is incremented once per number field](https://github.com/martijnvg/elasticsearch/commit/c714748e856ba4190e5f7ab5c1f25d73e3c41c0a#diff-aa342f9a6ef361b8871c6181dcc0ce36R321) when processing the percolator query at index time. I think that this is acceptable, percolator queries with range queries are more prone to false positives than percolator queries with term based queries to begin with, due to how the extracted ranges are stored.
